File: app/src/main/java/mega/privacy/android/app/VideoCompressor.java
package mega.privacy.android.app;

import android.content.Context;
import android.os.StatFs;

import java.io.File;
import java.util.List;

import mega.privacy.android.app.utils.conversion.VideoCompressionCallback;
import mega.privacy.android.domain.entity.SyncRecord;
import timber.log.Timber;

public class VideoCompressor extends VideoDownsampling {

    private List<SyncRecord> pendingList;

    private String outputRoot;

    private VideoCompressionCallback updater;

    private long totalInputSize;

    private long totalRead;

    private int totalCount;

    private int currentFileIndex;

    public void stop() {
        setRunning(false);
        Timber.d("Video compressor stopped");
    }

    public VideoCompressor(Context context, VideoCompressionCallback callback, int quality) {
        super(context);
        this.updater = callback;
        this.quality = quality;
    }

    public void setPendingList(List<SyncRecord> pendingList) {
        this.pendingList = pendingList;
        totalCount = pendingList.size();
        Timber.d("Total compression videos count is %s", totalCount);
        calculateTotalSize();
    }

    private void calculateTotalSize() {
        for (int i = 0; i < totalCount; i++) {
            totalInputSize += new File(pendingList.get(i).getLocalPath()).length();
        }
        Timber.d("Total compression size is %s", totalInputSize);
    }

    public long getTotalInputSize() {
        return this.totalInputSize;
    }

    public void setOutputRoot(String root) {
        this.outputRoot = root;
    }

    private boolean notEnoughSpace(long size) {
        double availableFreeSpace = Double.MAX_VALUE;
        try {
            StatFs stat = new StatFs(outputRoot);
            availableFreeSpace = stat.getAvailableBytes();
        } catch (Exception ex) {
            ex.printStackTrace();
        }
        return (size > availableFreeSpace);
    }

    public void start() {
        setRunning(true);
        for (int i = 0; i < totalCount && isRunning(); i++) {
            currentFileIndex = i + 1;
            SyncRecord record = pendingList.get(i);
            Timber.d("Video compressor start: %s", record.toString());
            String path = record.getLocalPath();
            File src = new File(path);
            long size = src.length();
            if (notEnoughSpace(size)) {
                updater.onInsufficientSpace();
                return;
            }

            VideoUpload video = new VideoUpload(path, record.getNewPath(), size, -1);
            try {
                prepareAndChangeResolution(video);
                if (isRunning()) {
                    updater.onCompressSuccessful(record);
                }
            } catch (Exception ex) {
                Timber.e(ex);
                updater.onCompressFailed(record);
                currentFileIndex++;
                totalRead += size;
            }
        }
        updater.onCompressFinished(totalCount + "/" + totalCount);
        stop();
    }

    public int getTotalCount() {
        return totalCount;
    }

    public int getCurrentFileIndex() {
        return currentFileIndex;
    }

    public long getTotalRead() {
        return totalRead;
    }
}

File: app/src/main/java/mega/privacy/android/app/VideoCompressor.kt
package mega.privacy.android.app

import android.content.Context
import android.os.StatFs
import mega.privacy.android.app.utils.conversion.VideoCompressionCallback
import mega.privacy.android.domain.entity.SyncRecord
import timber.log.Timber
import java.io.File

/**
 * The class to compress videos
 *
 * @property context The [Context] object
 * @property callback The [VideoCompressionCallback]
 * @property videoQuality The video quality to compress to
 */
class VideoCompressor(
    context: Context,
    private val callback: VideoCompressionCallback,
    videoQuality: Int,
) : VideoDownSampling(
    context = context,
    videoQuality = videoQuality,
) {
    private var pendingList = listOf<SyncRecord>()
    private var outputRoot: String? = null
    private var totalRead = 0L

    /**
     * Total input size
     */
    var totalInputSize = 0L

    /**
     * Overall video count
     */
    var totalCount = 0

    /**
     * Index of current file
     */
    var currentFileIndex = 0

    /**
     * Stops the video compression
     */
    fun stop() {
        setVideoDownSamplingRunning(false)
        Timber.d("Video compressor stopped")
    }

    /**
     * Sets the list of videos to be compressed
     *
     * @param pendingList A list of [SyncRecord]
     */
    fun setPendingList(pendingList: List<SyncRecord>) {
        this.pendingList = pendingList
        totalCount = pendingList.size
        Timber.d("Total compression videos count is %s", totalCount)
        calculateTotalSize()
    }

    /**
     * Calculates the total size of all videos to be compressed
     */
    private fun calculateTotalSize() {
        for (i in 0 until totalCount) {
            pendingList[i].localPath?.let { nonNullLocalPath ->
                totalInputSize += File(nonNullLocalPath).length()
            }
        }
        Timber.d("Total compression size is %s", totalInputSize)
    }

    /**
     * Sets the Output Root [String]
     *
     * @param root The Root [String], which can be nullable
     */
    fun setOutputRoot(root: String?) {
        outputRoot = root
    }

    /**
     * Checks whether there is enough space to compress the videos
     *
     * @return Whether there is enough space to compress the videos or not
     */
    private fun notEnoughSpace(size: Long): Boolean {
        val availableFreeSpace = try {
            StatFs(outputRoot).availableBytes.toDouble()
        } catch (exception: Exception) {
            Timber.e("Exception When Retrieving Free Space: $exception")
            Double.MAX_VALUE
        }
        return size > availableFreeSpace
    }

    /**
     * Starts the video compression
     */
    fun start() {
        setVideoDownSamplingRunning(true)
        var i = 0
        while (i < totalCount && isVideoDownSamplingRunning()) {
            currentFileIndex = i + 1
            val record = pendingList[i]
            Timber.d("Video compressor start: %s", record.toString())
            val path = record.localPath

            path?.let { nonNullPath ->
                val src = File(nonNullPath)
                val size = src.length()
                if (notEnoughSpace(size)) {
                    callback.onInsufficientSpace()
                    return
                }

                val newPath = record.newPath
                if (newPath == null) {
                    callback.onCompressFailed(record)
                    return
                }

                val video = VideoUpload(
                    originalPath = path,
                    newPath = newPath,
                    size = size,
                    pendingMessageId = -1,
                )
                try {
                    prepareAndChangeResolution(video)
                    if (isVideoDownSamplingRunning()) {
                        callback.onCompressSuccessful(record)
                    }
                } catch (ex: Exception) {
                    Timber.e(ex)
                    callback.onCompressFailed(record)
                    currentFileIndex++
                    totalRead += size
                }
            }
            i++
        }
        callback.onCompressFinished("$totalCount/$totalCount")
        stop()
    }
}
